src/main/java/uk/gov/pay/publicauth/service/DatabaseMetricsService.java

```diff
@@ -0,0 +1,75 @@
+package uk.gov.pay.publicauth.service;
+
+import com.codahale.metrics.Gauge;
+import com.codahale.metrics.MetricRegistry;
+import io.dropwizard.db.DataSourceFactory;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.Set;
+
+import static java.lang.String.format;
+
+public class DatabaseMetricsService {
+
+    private static final Set<PostgresMetric> metrics = Set.of(
+            new LongPostgresMetric("numbackends", 0L),
+            new LongPostgresMetric("xact_commit", 0L),
+            new LongPostgresMetric("xact_rollback", 0L),
+            new LongPostgresMetric("blks_read", 0L),
+            new LongPostgresMetric("blks_hit", 0L),
+            new LongPostgresMetric("tup_returned", 0L),
+            new LongPostgresMetric("tup_fetched", 0L),
+            new LongPostgresMetric("tup_inserted", 0L),
+            new LongPostgresMetric("tup_updated", 0L),
+            new LongPostgresMetric("tup_deleted", 0L),
+            new LongPostgresMetric("conflicts", 0L),
+            new LongPostgresMetric("temp_files", 0L),
+            new LongPostgresMetric("temp_bytes", 0L),
+            new LongPostgresMetric("deadlocks", 0L),
+            new DoublePostgresMetric("blk_read_time", 0.0),
+            new DoublePostgresMetric("blk_write_time", 0.0));
+
+    private final String databaseName;
+    private final DataSourceFactory dataSourceFactory;
+    private Integer statsHealthy = 0;
+
+    public DatabaseMetricsService(DataSourceFactory dataSourceFactory, MetricRegistry metricRegistry, String databaseName) {
+        this.dataSourceFactory = dataSourceFactory;
+        this.databaseName = databaseName;
+        initialiseMetrics(metricRegistry);
+    }
+
+    private void initialiseMetrics(MetricRegistry metricRegistry) {
+        metrics.forEach(metric -> metric.register(metricRegistry, format("%sdb.", databaseName)));
+        metricRegistry.<Gauge<Integer>>register(format("%sdb.stats_healthy", databaseName), () -> statsHealthy);
+    }
+
+    public void updateMetricData() {
+        try (Connection connection = DriverManager.getConnection(
+                dataSourceFactory.getUrl(),
+                dataSourceFactory.getUser(),
+                dataSourceFactory.getPassword())) {
+            connection.setReadOnly(true);
+            try (PreparedStatement statement = connection.prepareStatement("select * from pg_stat_database where datname = ?")) {
+                statement.setString(1, databaseName);
+                if (statement.execute()) {
+                    try (ResultSet resultSet = statement.getResultSet()) {
+                        if (resultSet.next()) {
+                            for (PostgresMetric metric : metrics) {
+                                metric.setValueFromResultSet(resultSet);
+                            }
```

Comment:
I was going to suggest using `metrics.forEach(metric -> metric.setValueFromResultSet(resultSet))` here but of course we can’t because it might throw a checked `SQLException`!

Reply:
Yes, the JDBC API is not very lambda-friendly